src/gluonts/nursery/ts_embeddings/embed_model.py

```diff
@@ -0,0 +1,329 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+
+from typing import Optional, List, Any
+import re
+import ast
+from argparse import ArgumentParser, ArgumentError, ArgumentTypeError
+
+import numpy as np
+import torch
+import torch.nn.functional as F
+
+import pytorch_lightning as pl
+
+from . import causal_cnn
+from . import contrastive_loss
+from .lars_wrapper import LARSWrapper
+
+import logging
+
+logging.getLogger("lightning").setLevel(logging.INFO)
+
+
+def try_add_arg(parser, *args, **kwargs):
+    try:
+        parser.add_argument(*args, **kwargs)
+    except ArgumentError:
+        logger = logging.getLogger(__name__)
+        logger.warning(f"error adding arguments args={args}, kwargs={kwargs}")
+
+
+def str2bool(v):
+    if isinstance(v, bool):
+        return v
+    if v.lower() in ("yes", "true", "t", "y", "1"):
+        return True
+    elif v.lower() in ("no", "false", "f", "n", "0"):
+        return False
+    else:
+        raise ArgumentTypeError("Boolean value expected.")
+
+
+class PreProcessor:
+    """
+    A PreProcessor can be used before the embedding model is used.
+    The `num_out_dim` indicates the number of outputs (e.g. 1 if the pre-processor returns scalar time series).
+    This is later used as input channels in the convolution layer.
+    The pre-processor should only use torch operations, since it will be stored as part of the final embedding model.
+
+    Note: the preprocessor operates on a batch, not the entire data set.
+    """
+
+    num_out_dim: int
+
+    def __call__(self, ts_batch):
+        """
+        Pre-processes a batch of time series with equal length.
+        The input shape should be
+        (N, T) or (N, C, T)
+
+        The output shape should be
+
+        (N, num_out_dim, L)
+        """
+        raise NotImplementedError()
+
+
+class ScalePreProcessor:
+    """
+    A pre-processor that takes a (possibly multi-variate) time series and normalizes each dimension by its mean
+    """
+
+    def __init__(self, out_dim, min_scale=1e-12):
```

Comment:
Scaling by a very small scale can have weird side effects (blowing up of values). Maybe better to not scale at all in this case?